`@simd` breaks generic code

vchuravy opened the pull request:
`@simd` has the assumption that are no loop dependencies (w.r.t to memory accesses) and that
can break generic code (see #27482), this removes that assumption from `@simd` (and therefore makes it less useful),
but still keeps its properties of allowing it to reorder floating point calculations (especially reductions). 
Since the assumptionof no loop dependencies can be valid (and is what Clang does for OpenMP) this PR
introduces a unsafe `@simdivdep` macro that retains that assumption.

@mbauman your docs changes from #27495 and the test for BitArrays should probably be part of this.
(Feel free to push to this branch :))

I am also open for renaming `@simdivdep` into something less horrible, but maybe this "feature"
deserves this name. Eventually we need to provide better aliasing information and fix the non-local
nature of `@simd` that Matt rightfully noticed in his PR.

vchuravy committed:
Don't assume parallel memory accesses in simd macro

If your code can gurantee that parallel memory accesses are legal you can
use `@simd ivdep` to give the compiler extra leeway. For generic code this
is an unsafe assumption.

Co-authored-by: Matt Bauman <mbauman@gmail.com>

vtjnash commented on src/llvm-simdloop.cpp:
```diff
@@ -172,14 +172,18 @@ void LowerSIMDLoop::enableUnsafeAlgebraIfReduction(PHINode *Phi, Loop *L) const
+        marker = M.getFunction("julia.simdivdep_marker");
+        if (!marker)
+            return false;
+    }
```
So you only get to use one of these options per Module? :P

vchuravy replied:
Oh that's what this is doing :P I thought that at first but then I convinced myself otherwise.
So I should look at `marker->users()`... I didn't want to duplicate that code.

vtjnash replied:
>  I didn't want to duplicate that code.

If only there was a way to write code once and then run it with two different values...

vtjnash reviewed (approved):
But seems like `@simd ivdep` is unused?

JeffBezanson commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27670#issuecomment-398567909) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/bd8d53ed0c26003bc37977d6e62acac643714b81/a984951_vs_333ea8c/report.md). cc @ararslan

martinholters commented:
How does this reduce allocations? :thinking:

vchuravy commented:
Or improve performance?...

On Wed, Jun 20, 2018, 00:40 Martin Holters <notifications@github.com> wrote:

> How does this reduce allocations? 🤔
>
> —
> You are receiving this because you authored the thread.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/27670#issuecomment-398655013>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI3atL2-h8t9xRuFSJYbKSw0QsaWuKEks5t-fyDgaJpZM4UuQgp>
> .
>

KristofferC commented:
https://github.com/JuliaLang/julia/pull/27495#issuecomment-396806558

I've seen "unexplainable" results here and there for basically as long as we have had Nanosoldier. I think something is weird in that it isn't just master vs merge commit at the exact  time the command was run that is run.

mbauman commented:
Yeah, they are comparing master at different points.  This PR got the extra advantage of d3c1ae388e97e438aa46b96c4b1ba353ab36c104 and 2df233ad72717b06cb0727b807da699c02ba71e7, neither of which were in the comparison (at 333ea8c3f2c3f4d858c4b431ca6c63fc1ffc3d48)

JeffBezanson commented:
This seems to kill even float axpy performance. Is there any way we can get that back?

mbauman committed:
Document `@simd`

Copies the information from the performance tips into a docstring; makes it a bit scarier.

Co-authored-by: Valentin Churavy <v.churavy@gmail.com>

vchuravy commented:
IIUC the reason why axpy performance is bad is because LLVM can no longer assume that input and output are not aliasing, so something like #20257 should help with that.
But yes those performance degredations is the reason why I still want to have `@simd ivdep`.

mbauman commented:
Does this really need NEWS?

mbauman commented on base/simdloop.jl:
```diff
@@ -85,6 +85,36 @@ function compile(x, ivdep)
+The object iterated over in a `@simd for` loop should be a one-dimensional range or a CartesianIndices iterator.
+By using `@simd`, you are asserting several properties of the loop:
+
+    * It is safe to execute iterations in arbitrary or overlapping order, with special consideration for reduction variables.
```
So is it safe to remove this bullet point now?

vchuravy replied:
No I think that is still a property of `@simd` w.r.t to floating-point operations

vchuravy commented:
@nanosoldier `runbenchmarks("array" || "problem" || "random", vs="665b1def2710f1f5d956b47c670034be0600ebb1")`

mbauman commented on base/reducedim.jl:
```diff
@@ -230,6 +230,13 @@ function _mapreducedim!(f, op, R::AbstractArray, A::AbstractArray)
             end
             R[i1,IR] = r
         end
+    elseif is_simd_safe(R) # TODO make sure that R and A are not aliased?
```
```julia
elseif is_simd_safe(R) && !mightalias(R, A)
```

vchuravy commented:
@nanosoldier runbenchmarks("array" || "problem" || "random", vs="@665b1def2710f1f5d956b47c670034be0600ebb1")

JeffBezanson commented on base/broadcast.jl:
```diff
@@ -815,6 +816,8 @@ preprocess(dest, x) = extrude(broadcast_unalias(dest, x))
 preprocess_args(dest, args::Tuple{Any}) = (preprocess(dest, args[1]),)
 preprocess_args(dest, args::Tuple{}) = ()
 
+is_simd_safe(bc::Broadcasted) = all(is_simd_safe.(bc.args))
```
`all(is_simd_safe, bc.args)`

vtjnash replied:
This definition doesn't make sense to me. Either we should assume that `broadcast` is a read-only operation (and this function thus would always be `true`), or we can't assume that about the function, in which case we must always say `false`.

KristofferC commented:
@nanosoldier `runbenchmarks("array" || "problem" || "random", vs="@665b1def2710f1f5d956b47c670034be0600ebb1")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27670#issuecomment-399050652) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/fad7a2cc09a2492bd32d894e5df1448e839eb243/835622a_vs_665b1de/report.md). cc @ararslan

vtjnash commented on base/broadcast.jl:
```diff
@@ -826,8 +829,14 @@ preprocess_args(dest, args::Tuple{}) = ()
     bc′ = preprocess(dest, bc)
-    @simd for I in eachindex(bc′)
-        @inbounds dest[I] = bc′[I]
+    if is_simd_safe(dest) && is_simd_safe(bc′)
```
`bc` is read-only memory – I don't think it matters whether it is safe to write to it

mbauman commented:
Yes I agree with Jameson's now-hidden comment (due to my poorly-timed commit).  https://github.com/JuliaLang/julia/pull/27670#discussion_r197222027

> Either we should assume that `broadcast` is a read-only operation (and this function thus would always be `true`), or we can't assume that about the function, in which case we must always say `false`.

Whereas my PR attempted to green-light _functions_, this is green-lighting _arguments_.  Neither is sufficient.

mbauman commented:
We just need to get _some_solution in here now to get the beta out the door.  I propose we remove these `is_simd_safe` workarounds and live with the perf hit for now.  I think we'll need to individually green-light entire method signatures, with things like:

```
is_simd_safe(::typeof(+), ::Float64, ::Float64...) = true
```

Keno commented:
Alright folks, this has gone on long enough. Let's get something figured out here today, or I'll disable `@simd` completely for the beta.

mbauman commented:
Let's try to get a clean nano soldier run through without the workarounds:

@nanosoldier `runbenchmarks("array" || "problem" || "random", vs="@665b1def2710f1f5d956b47c670034be0600ebb1")`

mbauman commented:
Well, here I thought reverting the somewhat piece-meal workarounds here would make this easier to get merged, but it turns out that's not the case.  We discussed this at the beginning of the triage call.  Jeff was adamantly against merging this with the 8x performance regressions in axpy, which is simply testing:

```julia
julia> function perf_axpy!(a, X, Y)
           @simd for i in eachindex(X)
               @inbounds Y[i] += a*X[i]
           end
           return Y
       end

julia> X = rand(Float32, 4096)
       a = rand(Float32);

julia> @btime perf_axpy!($a, $X, $X);
  2.750 μs (0 allocations: 0 bytes) # On master I get ~290 ns
```

Now, what's fascinating here is that _this performance penalty only applies when `X` aliases `Y`_!

```julia
julia> @btime perf_axpy!($a, $X, $Y);
  291.964 ns (0 allocations: 0 bytes) # Matches master performance
```

Note that this occurs _automatically_, with or _without_ and explicit `@simd`!  This is really cool.  My LLVM reading skills are a little rough, but it appears as though LLVM is packaging both scalar and vector code here, and dynamically branching to them depending on the overlap of the `addrspace`.

So, given that information, does that assuage your concerns here, @JeffBezanson?

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/27670#issuecomment-399199146):
```
NanosoldierError: failed to run benchmarks against comparison commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3aa616d2511b7f2cbea533d53dcf1423aa28a660/896176e_vs_665b1de/report.md)
cc @ararslan

JeffBezanson commented:
Yes, that's excellent. In that case we can proceed with this. We should add benchmark cases both with and without aliasing.

vchuravy commented:
What is the conclusion here? Jameson's concern was about merging `@simd ivdep` without a use/test-case.

Matt, while I understand your concern about algebraic reordering, that can be a statement of a generic algorithm so I still don't see why you would want to mark operations safe.
The primary concern of this PR is an attribute of memory operations on datatypes, and so my version of `is_simd_safe` (still not a great name I know) states that memory operations on
this data type are safe to be done in parallel. 

So shall we remove the last two commits from this and merge or do we decide not to be too clever and don't do `@simd ivdep`.

mbauman commented:
Well, we definitely cannot mark broadcast `@simd ivdep` without any consideration of the methods involved.  It's such a generic construct, and we allow broadcasting _everywhere_.  For example, the following function would fail your case:

```julia
const bitcache = falses(64);
f(v, i) = (bitcache[i] = xor(bitcache[i], v))
R = Array{Bool}(4)
R .= f.(true, [1,2,3,4])
```

I think it's fine and reasonable to allow `@simd` to continue having an `ivdep` expert behavior — as we've seen above it's necessary in the case where the arrays alias but the author knows that the reads/writes from successive iterations are independent.

Can't we just add a testcase for `@simd ivdep` and simply ensure that it still performs the loop?

vchuravy commented:
OOM on AV 64bit during sysimg built

vchuravy commented:
Travis is: ` Got exception UDP send failed: address not available (EADDRNOTAVAIL) outside of a @test`
